### ISSUE TITLE
SC1035

### ISSUE THREAD

**Issue by luizluca** [koalaman/shellcheck#1280]
#### For bugs
- Rule Id SC1035
- My shellcheck version: online
- [X] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash
[_version() {
        true
}

```

#### Here's what shellcheck currently says:
Line 1:
[_version() {
^-- SC1035: You need a space after the [ and before the ].
^-- SC1073: Couldn't parse this test expression. Fix to allow more checks.
         ^-- SC1036: '(' is invalid here. Did you forget to escape it?
         ^-- SC1072: Expected test to end here (don't wrap commands in []/[[]]). Fix any mentioned problems and try again.

#### Here's what I wanted or expected to see:
No error

**Comment by ngzhian** [comment]
function names cannot contain brackets, i.e. [

**Comment by matthewpersico** [comment]
Apparently, they can:

```bash
$ bash --version
GNU bash, version 4.3.42(1)-release (x86_64-unknown-linux-gnu)
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

$ [_version() {         echo hola; }

$ [_version                          
hola

 $ _version
-bash: _version: command not found
```

**Comment by matthewpersico** [comment]
But this seems one-off, must be the first char or something else weird is going on:

```bash
$ _vers[ion() {         echo hola; }
> ]
-bash: $'_vers[ion() {         echo hola; }\n]': command not found
```

**Comment by luizluca** [comment]
POSIX does not prohibit different characters as "An implementation may allow other characters in a function name as an extension"

FYI, [_version() is an alternative test for version numbers, to use like this:

```sh
if [_version $MYAPP_VERSION -gt 4 ] && [_version $MYAPP_VERSION -lt 4.4.12.3 ]; then
   do_something
fi
```

**Comment by matthewpersico** [comment]
Can you please explain where I've gone wrong interpreting your intention and/or using your code:

```bash
$ MYAPP_VERSION=4.3
$ if [_version $MYAPP_VERSION -gt 4 ] && [_version $MYAPP_VERSION -lt 4.4.12.3 ]; then echo "MYAPP_VERSION ($MYAPP_VERSION) is in the middle"; fi 
MYAPP_VERSION (4.3) is in the middle

$ MYAPP_VERSION=3
$ if [_version $MYAPP_VERSION -gt 4 ] && [_version $MYAPP_VERSION -lt 4.4.12.3 ]; then echo "MYAPP_VERSION ($MYAPP_VERSION) is in the middle"; fi 
MYAPP_VERSION (3) is in the middle

$ MYAPP_VERSION=5
$ if [_version $MYAPP_VERSION -gt 4 ] && [_version $MYAPP_VERSION -lt 4.4.12.3 ]; then echo "MYAPP_VERSION ($MYAPP_VERSION) is in the middle"; fi 
MYAPP_VERSION (5) is in the middle
``

**Comment by loren-osborn** [comment]
From an outside observer:

This seems to be a case of highly idiomatic programming. If this is a common idiom, I see no reason it shouldn’t be supported, but it isn’t a bash idiom i’ve seen before. Many languages tend to call code similar to this “magic”. If it is an uncommon idiom, it might be worth handling, but printing a warning to ensure it isn’t misused... probably also add a check for expressions that match a function name like this, except a space is accidentally typed instead of an underscore?

Just my totally random 2¢.

**Comment by loren-osborn** [comment]
https://goo.gl/images/QpXh2c

**Comment by luizluca** [comment]
"[" is just the same as "test". The only difference I know is that "[" requires "]" as the last command line argument. In bash, it is implemented as builtin command but remember that '/usr/bin/[' does exist. 
Even on a shell that does not accept function names starting with '[', you can:

```sh
$ [xxx() { a; }
sh: syntax error: bad function name
$ echo '[ "$@"' > /usr/bin/[xxx
$ chmod +x /usr/bin/[xxx
$ if [xxx 1 = 1 ]; then echo yes; else echo no; fi
yes
```

Bash "parser" (or any other shell) does not really check anything related to "[". It only cares that everything after an "if" or "while" is a valid command (with valid arguments). It is "[" command that demand "]", not bash parser. It is not different from:

```sh
_version() {
    (...)
}
if _version $MYAPP_VERSION -gt 4 ] && _version $MYAPP_VERSION -lt 4.4.12.3 ]; then echo "MYAPP_VERSION ($MYAPP_VERSION) is in the middle 1"; fi
```

The simple question is: would [_xxx be an acceptable function name for shellcheck (as it is for bash)? 

BTW, the "underscore" is just any character. "[xxx" is also accepted. No need to test as a special case.
